src/ast/passes/semantic_analyser.cpp

```diff
@@ -514,7 +519,51 @@ void SemanticAnalyser::visit(Call &call)
       auto &expr = (*call.vargs)[i];
       func_arg_idx_ = i;
 
-      expr->accept(*this);
+      // Handling offsetof() 2nd argument
+      // Resolve the second offsetof() parameter as early as possible,
+      // otherwise, we need to do too much adaptation work in the semantic
+      // analyzer stage. For example:
+      //
+      //  struct foo {
+      //    int offsetof;   > offsetof(struct foo, offsetof)
+      //    int kstack;     > offsetof(struct foo, kstack)
+      //    int ctx;        > offsetof(struct foo, ctx)
+      //    int comm;       > offsetof(struct foo, comm)
+      //  }
+      if (func_ == "offsetof" && func_arg_idx_ == 1)
+      {
+        auto fc = dynamic_cast<FieldAccess *>(call.vargs->at(1));
+        auto binop = dynamic_cast<Binop *>(call.vargs->at(1));
+
+        // TODO: Only support single-member form for now.
+        // a.b in offsetof(struct Foo, a.b) is illegal
+        if (fc != nullptr)
+        {
+          LOG(ERROR, call.loc, err_)
+              << call.func << "() unsupport subfield format.";
+        }
+        // a+b in offsetof(struct Foo, a+b) is illegal
+        else if (binop != nullptr)
+        {
+          LOG(ERROR, call.loc, err_)
+              << call.func << "() unsupport expression format.";
+        }
```

Comment:
It looks like you're checking for very specific invalid usage. Is that necessary? Maybe it'd be simpler if we blindly coerce the second argument to an identifier. And then later when `offsetof` is processed below it'll just error and say something like "struct foo has no field 'a+b'". That error seems good enough and I can't really imagine anyone thinking `offsetoff(struct task_struct, foo+bar)` being  valid usage.

Reply:
`offsetof(struct task_struct, comm)`'s `comm` will `dynamic_cast` to `nullptr`, cause `comm` is a builtin. same as `offsetof(struct Foo, offsetof)`'s `offsetof`, `offsetof(struct Foo, bswap)`'s `bswap`. I don't find a better way to fix these situation.

Reply:
I just force push a resolve, please review.